### PR TITLE
Add support for yoga context

### PR DESCRIPTION
**Pull request by victorporof**
Depends on https://github.com/bschwind/yoga-rs/pull/26

### CODE REVIEW

**Comment by victorporof** [comment]
Rebased.

**Comment by victorporof** [comment]
Added another test.

**Comment by bschwind** [comment]
I'm taking a look, should have a review for you fairly soon.

**Comment by victorporof** [comment]
@bschwind Take your time. I'm just noting any updates I'm adding, not pinging you for a quicker review. Thanks!

File: src/lib.rs [hunk 1]
```diff
@@ -1483,14 +1461,39 @@ impl Node {
...
+		unsafe { &mut *(internal::YGNodeGetContext(node.inner_node) as *mut Context) }
+	}
+
+	pub fn get_context_raw(node_ref: &NodeRef) -> &mut Context {
+		unsafe { &mut *(internal::YGNodeGetContext(*node_ref) as *mut Context) }
+	}
+}
+
+#[derive(Debug)]
+pub struct Context(pub Box<Any>);
```
Comment (bschwind) [inline src/lib.rs]: Is it possible to use generics here? I'm not too knowledgeable on C FFI when dealing with the Any type vs. using generics

Comment (victorporof) [reply src/lib.rs]: I think so! I'll get back to you in a bit.

**Comment by victorporof** [comment]
Updated PR using generics.

**Comment by bschwind** [comment]
Thanks for the update! What's your opinion on the generics version vs. your original implementation? I just want to make sure the API is convenient as possible. The generics one looks a bit nicer to use but if you have strong opinions one way or the other, let me know!

**Comment by victorporof** [comment]
I'm ambivalent to both, they're both awful. Callbacks through FFI are a horrible thing.

It's going to be awkward no matter what we do, since getting context back from an internal yoga node involves an unsafe pointer coercion to a reference, and we need to do that without having any knowledge of the original type (since the callback is passed a NodeRef, not a Node). This means that there's no way of knowing the type of the original context inside the callback. This is unlikely to be an issue in practice, since I don't see a real need for using different types as contexts for different nodes, but who knows?

The implementation using generics is indeed cleaner, and avoids a lot of noise versus the Box<Any> variant. The problem with this is that calls to `get_context` using an incorrect type result in undefined behaviour, since you're converting a raw pointer to a type which might not be the one you started with. As I've said, I don't see this an issue in practice, but it's a point of failure.

On the other hand, the implementation using Box<Any> is much noisier and has all of the disadvantages of using Any, but doesn't result in undefined behaviour, since we need to go through `downcast_ref` on the Box type, which returns an option that we can explicitly handle. If you pass in an incorrect type, you get None, instead of the undefined behaviour like with generics.

A third option would be to avoid generics and Any types altogether, and create a custom context type which we hope handles most use cases. I don't see how we could do this. I'd like font data and strings inside contexts for measure functions, but other people might want something else.

**Comment by victorporof** [comment]
I'll add a commit to this PR switching things back to using Box<Any>. I'll let you decide which one you prefer (lots of API noise but no undefined behaviour, vs. clean API but potential undefined behaviour if you don't know which type to coerce back to when getting context)

**Comment by victorporof** [comment]
See 6469380

**Comment by victorporof** [comment]
The more I think about it, the more I'm leaning towards the "lesser of two evils", which is using Box<Any>. I this case at the very least there's no chance of passing in an incorrect type and mysteriously succeed when retrieving context.

**Comment by bschwind** [comment]
Yeah I agree, I'd opt to take the path that results in no undefined behavior. Thanks for laying it out and describing where UB can happen. Let's go with the Box solution. This functionality probably won't be used too much except by people who are doing lower level stuff, so they can refer to the tests if the API isn't clear.

**Comment by victorporof** [comment]
@bschwind Yeah, that's why I added multiple tests. Feel free to merge this PR when ready, I have no other changes to make.

**Comment by bschwind** [comment]
Looks good to me, merging. Thanks for the PR, this is good functionality to have!

**Pull request merged**